src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java

```diff
@@ -1,31 +1,272 @@
+                    ArtifactFilter filter = null; // Evaluate all scopes
+                    node = dependencyCollectorBuilder.collectDependencyGraph(buildingRequest, filter);
+                } catch (DependencyCollectorBuilderException e) {
+                    throw new MojoExecutionException("Failed to analyze dependency tree for useUpperBounds", e);
+                }
+                RequireUpperBoundDepsVisitor visitor = new RequireUpperBoundDepsVisitor();
+                node.accept(visitor);
+                Map<String, String> upperBounds = visitor.upperBounds();
+
+                if (!upperBounds.isEmpty()) {
+                    // Second pass: apply the results of the upper bounds analysis.
+
+                    /*
+                     * applyOverrides depends on resolution, so resolve again between the first pass
+                     * and the second.
+                     */
+                    Set<Artifact> resolved = resolveDependencies(shadow);
+                    shadow.setArtifacts(resolved);
+
+                    // We intentionally want to use a lower version of this dependency.
+                    String servletHack = upperBounds.get("javax.servlet:servlet-api");
+                    if (servletHack != null && !servletHack.equals("0")) {
+                        upperBounds.remove("javax.servlet:servlet-api");
+                    }
+
+                    applyOverrides(upperBounds, Collections.emptyMap(), shadow, getLog());
+                }
+            }
+
+            /*
+             * At this point, the model has been updated as the user has requested. We now redo
+             * resolution and compare the new resolution to the original in order to account for
+             * updates to transitive dependencies that are not present in the model. Anything that
+             * was updated in the new resolution needs to be updated in the test classpath. Anything
+             * that was removed in the new resolution needs to be removed from the test classpath.
+             * Anything that was added in the new resolution needs to be added to the test
+             * classpath.
+             */
+            Set<Artifact> resolved = resolveDependencies(shadow);
+            effectiveArtifacts = wrap(new Artifacts(resolved));
+            Map<String, String> newResolution = new HashMap<>();
+            for (Artifact artifact : resolved) {
+                newResolution.put(toKey(artifact), artifact.getVersion());
+            }
+            for (Map.Entry<String, String> entry : newResolution.entrySet()) {
+                if (originalResolution.containsKey(entry.getKey())) {
+                    // Present in both old and new resolution: check for update.
+                    String originalVersion = originalResolution.get(entry.getKey());
+                    String newVersion = entry.getValue();
+                    if (!newVersion.equals(originalVersion)) {
```

Comment:
should this not be only doing this if the version is newer (using a maven version comparison) and presumably aborting if the war contains an older version of the dependency?
(in other words the prepared war may have all the latest releases of a plugin, but if you are running this on a plugin that has a dependency on an incremenatal the results will be incorrect and this user error should be reported rather than ignored?)

Reply:
I thought so too originally, but that line of reasoning turned out to be flawed. The situation is as follows: suppose we depend on A:1.0 which depends on B:1.2. Now suppose a problem is discovered in B:1.2 that results in A:1.1 rolling back to B:1.1. We only ever directly depended on A:1.0, but now we override A:1.0 to A:1.1. B:1.2 was in our transitive tree before, but now for correctness we must change B from 1.2 to 1.1.

Reply:
Makes sense thanks for the explanation .  Worth a comment for future contributors?

```suggestion
                    /*
                     * We do an equals check here rather than a new for the following reason: 
                     *  suppose we depend on A:1.0 which depends on B:1.2. Now suppose a problem is discovered in B:1.2
                     * that results in A:1.1 rolling back to B:1.1. We only ever directly depended on A:1.0, but now we
                     *  override A:1.0 to A:1.1. B:1.2 was in our transitive tree before, but now for correctness we must
                     * change B from 1.2 to 1.1.
                     */
                    if (!newVersion.equals(originalVersion)) {
```